File: js/diamond.js
// JavaScript uses mixins/composition instead
class A { foo() { console.log("A"); } }
class B extends A { foo() { console.log("B"); } }
class C extends A { foo() { console.log("C"); } }

// Composition-based workaround
class D {
  constructor() {
    this.b = new B();
    this.c = new C();
  }
  foo() { this.b.foo(); } // Choose implementation explicitly
}

const d = new D();
d.foo(); // Output: B

File: scala/src/diamond.scala
trait A { def foo(): Unit = println("A") }
trait B extends A { override def foo(): Unit = println("B") }
trait C extends A { override def foo(): Unit = println("C") }

class D extends B with C // C's implementation takes precedence

object Main extends App {
  val d = new D
  d.foo() // Output: C
}